Lib/idlelib/formatregion.py

```diff
@@ -0,0 +1,178 @@
+"""Format a selected region of text.
+
+Several formatting options are available:
+indent, deindent, comment, uncomment, tabify, and untabify.
+"""
+from tkinter.simpledialog import askinteger
+
+
+def classifyws(s, tabwidth):
+    """Classify leading whitespace in a string.
+
+    Count the number of actual spaces and tabs at the beginning of
+    the string.  Also count the tabwidth number of spaces that a
+    tab would expand to.
+
+    Return a tuple of (number of raw leading whitespace characters,
+        number of leading characters after expanding tabs).
+    """
+    raw = effective = 0
+    for ch in s:
+        if ch == ' ':
+            raw = raw + 1
+            effective = effective + 1
+        elif ch == '\t':
+            raw = raw + 1
+            effective = (effective // tabwidth + 1) * tabwidth
+        else:
+            break
+    return raw, effective
+
+
+class FormatRegion:
+    "Format selected text."
+
+    def __init__(self, editwin):
+        self.editwin = editwin
+
+    def get_region(self):
+        """Return line information about the selected text region.
+
+        If text is selected, the first and last indices will be
+        for the selection.  If there is no text selected, the
+        indices will be the current cursor location.
+
+        Return a tuple containing (first index, last index,
+            string representation of text, list of text lines).
+        """
+        text = self.editwin.text
+        first, last = self.editwin.get_selection_indices()
+        if first and last:
+            head = text.index(first + " linestart")
+            tail = text.index(last + "-1c lineend +1c")
```

Comment:
I think this is missing a space... do the tests cover the line ending edge cases?
```suggestion
            tail = text.index(last + " -1c lineend +1c")
```